File: llm_outputs/python_programs_corrected_by_llama/lis.py
def lis(arr):
    ends = {i: float('-inf') for i in range(len(arr))}
    longest = 0

    for i, val in enumerate(arr):

        prefix_lengths = [j for j in range(longest + 1) if arr[ends[j]] < val]

        length = max(prefix_lengths) if prefix_lengths else 0

        if length > longest or (length == longest and val <= arr[ends[length]]):
            ends[i+1] = length
            longest = length

    return longest

File: llm_outputs/java_programs_corrected_by_llama/LIS.java
package java_programs;
import java.util.*;

public class LIS {
    public static int lis(int[] arr) {
        Map<Integer, Integer> ends = new HashMap<>();
        int longest = 0;

        int i = 0;
        for (int val : arr) {

            ArrayList<Integer> prefix_lengths = new ArrayList<>();
            for (Map.Entry<Integer, Integer> entry : ends.entrySet()) {
                if (entry.getValue().equals(i)) continue;
                if (arr[entry.getKey()] < val) {
                    prefix_lengths.add(entry.getKey());
                } else break;
            }

            int length =!prefix_lengths.isEmpty()? Collections.max(prefix_lengths) : 0;

            if ((longest <= length && val > arr.get(ends.get(longest))) ||
                (val >= arr.get(Math.min(ends.keySet(), longest)))) {
                ends.put(length, i);
                longest = Math.max(longest, length);
            }

            i++;
        }
        return longest;
    }
}